wire-ios-data-model/Source/E2EIdentity/MLSConversationVerificationManager.swift

```
@@ -0,0 +1,52 @@
+//
+// Wire
+// Copyright (C) 2023 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+public protocol MLSConversationVerificationManagerInterface {
+
+    func startObservingMLSConversationVerificationStatus()
+
+}
+
+public final class MLSConversationVerificationManager: MLSConversationVerificationManagerInterface {
+
+    // MARK: - Properties
+
+    var mlsService: MLSServiceInterface
+    var mlsConversationVerificationStatusProvider: MLSConversationVerificationStatusProviderInterface?
+
+    // MARK: - Life cycle
+
+    public init(mlsService: MLSServiceInterface,
+                mlsConversationVerificationStatusProvider: MLSConversationVerificationStatusProviderInterface?) {
+        self.mlsService = mlsService
+        self.mlsConversationVerificationStatusProvider = mlsConversationVerificationStatusProvider
+    }
+
+    // MARK: - Methods
+
+    public func startObservingMLSConversationVerificationStatus() {
+        Task {
+            for try await groupID in mlsService.epochChanges() {
+                try await mlsConversationVerificationStatusProvider?.updateStatus(groupID)
+            }
```

1) if it fails the error is swallowed here
2) can we run each update in parallel?

---

I agree, we must add `do/catch` here.
Also we must not lose the reference to the task, we need to be able to cancel it.

---

> 2. can we run each update in parallel?

This is an `AsyncStream`.

---

https://github.com/wireapp/wire-ios/pull/1029